composer/trainer/_deepspeed.py

```diff
@@ -75,25 +75,23 @@ def _ensure_no_optim_in_config(config: Dict[str, Any]):
 def _add_precision_config(config: Dict[str, Any], state: State):
     precision = state.precision
 
+    # Verify DeepSpeed config is consistent with state.precision if set
     ds_precision = None
     if 'fp16' in config and 'enabled' in config['fp16'] and config['fp16']['enabled']:
-        ds_precision = Precision.FP16
-    if 'bf16' in config and 'enabled' in config['bf16'] and config['bf16']['enabled']:
-        raise ValueError(('DeepSpeed is configured to use BFLOAT16, but this is unsupported by the '
-                          'Mosaic trainer.'))
-    if 'amp' in config and 'enabled' in config['amp'] and config['amp']['enabled']:
-        raise ValueError(('DeepSpeed is configured to use Apex AMP, but this is unsupported by the '
-                          'Mosaic trainer.'))
-
+        ds_precision = Precision.AMP_FP16
+    elif 'bf16' in config and 'enabled' in config['bf16'] and config['bf16']['enabled']:
+        ds_precision = Precision.AMP_BF16
+    elif 'amp' in config and 'enabled' in config['amp'] and config['amp']['enabled']:
+        ds_precision = Precision.AMP_FP16
     if ds_precision is not None and ds_precision != precision:
         raise ValueError((f'Provided DeepSpeed configuration specifies precision={ds_precision}, '
                           f'but the Mosaic trainer has been configured with precision={precision}.'))
 
-    if precision == Precision.FP16:
-        if 'fp16' not in config:
-            config['fp16'] = cast(Dict[str, Any], {'enabled': True})
-        fp16_config = config['fp16']
-        assert isinstance(fp16_config, dict)
+    # Set DeepSpeed config based on state.precision if not set
+    if precision == Precision.AMP_FP16 and 'fp16' not in config:
+        config['fp16'] = cast(Dict[str, Any], {'enabled': True})
```

Comment:
What is the difference between config['fp16'] and config['amp']? Unclear which one corresponds to precision == Precision.AMP_FP16.

Reply:
Same as above, this is in Deepspeed docs and not our settings. It's really confusing